cli/src/main/java/org/projectnessie/tools/catalog/migration/cli/BaseRegisterCommand.java

```diff
@@ -0,0 +1,337 @@
+              + "of tables that will be registered by running this.")
+  private boolean isDryRun;
+
+  @CommandLine.Option(
+      names = {"--disable-safety-prompts"},
+      description = "Optional configuration to disable safety prompts which needs console input.")
+  private boolean disablePrompts;
+
+  @CommandLine.Option(
+      names = {"--stacktrace"},
+      description =
+          "Optional configuration to enable capturing stacktrace in logs in case of failures.")
+  private boolean enableStackTrace;
+
+  private static final int BATCH_SIZE = 100;
+  public static final String FAILED_IDENTIFIERS_FILE = "failed_identifiers.txt";
+  public static final String FAILED_TO_DELETE_AT_SOURCE_FILE = "failed_to_delete_at_source.txt";
+  public static final String DRY_RUN_FILE = "dry_run_identifiers.txt";
+
+  private static final Logger consoleLog = LoggerFactory.getLogger("console-log");
+
+  public BaseRegisterCommand() {}
+
+  protected abstract CatalogMigrator catalogMigrator(
+      Catalog sourceCatalog, Catalog targetCatalog, boolean enableStackTrace);
+
+  protected abstract boolean canProceed(Catalog sourceCatalog);
+
+  protected abstract String operation();
+
+  protected abstract String operated();
+
+  protected abstract String operate();
+
+  @Override
+  public Integer call() {
+    Set<TableIdentifier> identifiers = Collections.emptySet();
+    String identifierRegEx = identifierOptions != null ? identifierOptions.identifiersRegEx : null;
+
+    if (identifierOptions != null) {
+      identifiers = identifierOptions.processIdentifiersInput();
+    }
+    checkAndWarnAboutIdentifiers(identifiers, identifierRegEx);
+
+    validateOutputDir();
+
+    Catalog sourceCatalog = sourceCatalogOptions.build();
+    consoleLog.info("Configured source catalog: {}", sourceCatalog.name());
+
+    Catalog targetCatalog = targetCatalogOptions.build();
+    consoleLog.info("Configured target catalog: {}", targetCatalog.name());
+
+    if (!isDryRun && !disablePrompts && !canProceed(sourceCatalog)) {
+      return 2;
+    }
+
+    CatalogMigrator catalogMigrator =
+        catalogMigrator(sourceCatalog, targetCatalog, enableStackTrace);
+
+    if (identifiers.isEmpty()) {
+      consoleLog.info("Identifying tables for {} ...", operation());
+      identifiers = catalogMigrator.getMatchingTableIdentifiers(identifierRegEx);
+      if (identifiers.isEmpty()) {
+        consoleLog.warn(
+            "No tables were identified for {}. Please check `catalog_migration.log` file for more info.",
+            operation());
+        return 2;
+      }
+    }
+
+    if (isDryRun) {
+      consoleLog.info("Dry run is completed.");
+      handleDryRunResult(identifiers);
+      return 0;
+    }
+
+    consoleLog.info("Identified {} tables for {}.", identifiers.size(), operation());
+
+    consoleLog.info("Started {} ...", operation());
+
+    CatalogMigrationResult result;
+    try {
+      Iterable<List<TableIdentifier>> identifierBatches =
```

Comment:
Yes. It is overkill. But the intermediate prints to the user are super important in this scenario to know the progress.
As we previously discussed, API must not print the intermediate progress (it has to be from the CLI). Hence, splitting the processing into batches so that CLI can provide progress info.

Reply:
The "new solution" is yet another of many unnecessary big collection instances and copying. Still a way too complicated approach to just print progress every N items.

Reply:
As mentioned above, `CatalogMigrator.registerTables()` is part of api module and `registerTables` in api module must not print the progress on the console log as it is a CLI functionality. 
As `registerTables` accept a list of identifiers, if I pass all the identifiers, I cannot keep track of progress from CLI module. 
Hence, virtually dividing into batches and passing batch by batch. So, CLI module can print the progress after each batch. 

So, I am not sure about "just print progress every N items".  
Did you mean instead of `registerTables` accepting all the Identifiers, have one more interface to accept single table identifiers for `registerTables`? So that caller can just call this API for each identifier and print progress every N time?